core/src/layers/correctness_check.rs

```diff
@@ -0,0 +1,602 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::{Debug, Formatter};
+use std::sync::Arc;
+
+use crate::raw::{
+    Access, AccessorInfo, Layer, LayeredAccess, OpBatch, OpCopy, OpCreateDir, OpDelete, OpList,
+    OpPresign, OpRead, OpRename, OpStat, OpWrite, Operation, RpBatch, RpCopy, RpCreateDir,
+    RpDelete, RpList, RpPresign, RpRead, RpRename, RpStat, RpWrite,
+};
+use crate::{Error, ErrorKind};
+
+/// Add a correctness capability check layer for every operation
+///
+/// Before performing any operations, we will first verify the operation and its critical arguments
+/// against the capability of the underlying service. If the operation or arguments is not supported,
+/// an error will be returned directly.
+///
+/// # Notes
+///
+/// OpenDAL applies this checker to every accessor by default, so users don't need to invoke it manually.
+/// this checker ensures the operation and its critical arguments, which might affect the correctness of
+/// the call, are supported by the underlying service.
+///
+/// for example, when calling `write_with_append`, but `append` is not supported by the underlying
+/// service, an `Unsupported` error is returned. without this check, undesired data may be written.
+#[derive(Default)]
+pub struct CorrectnessChecker;
```

Comment:
Please use `CorrectnessCheckLayer` as the name to ensure that all layer names end with `Layer`.